README.md
# arvanApi

## General info
Simple ArvanCloud API for python based on [ArvanCloud API](https://www.arvancloud.ir/docs/api/cdn/4.0)

## Requirements
- [python 3](https://www.python.org/downloads)
- [ArvanCloud API key](https://panel.arvancloud.ir/profile/machine-user)

## Installation
  ```console
  pip install --upgrade git+https://github.com/mheidari98/arvanApi.git
  ```

## Uninstall
  ```console
  pip uninstall arvanapi
  ```


## Usage
There are two ways of using this package. The easy way is to use the `arvan` script, which allows you to manage your domains and DNS records from the command line. The other way is to use the `Arvan` class in your python code.

A quick example:
```console
# Arvan cli CDN Manager
arvan
```
---

You can also use it programmatically, like this:
```python
from arvanApi import Arvan

api_key = 'dfbeaaf2-a653-47f7-ad42-caf052d4a2b0'  # this is fake UUID, put your api key here
arv = Arvan(api_key, debug=False)

domain = "myDomain.com"
arv.createDomain(domain)

myDomain = arv.getDomain(domain)
myDomain.changeSsl(ssl_status=True)

ipAddr = '127.0.0.1'  # this is fake ip address, put your server ip address here
dirId = myDomain.createDnsARecord('dir', ipAddr, cloud=False)
cdnId = myDomain.createDnsARecord('cdn', ipAddr, port=80, cloud=True, upstream_https='http')

for key, dns in myDomain.DNSs.items():
    print(dns)
```

## Methods
- [x] create Domain
- [x] get Domain
- [x] delete Domain
- [x] create DNS Record
- [x] get DNS Record
- [x] delete DNS Record
- [x] get SSL status
- [x] change SSL status


## Status
Project is: _in progress_

## License
[MIT](https://choosealicense.com/licenses/mit)

## Contact
Created by [@mheidari98](https://github.com/mheidari98)

## Support
If you like this project, please consider supporting it by donating to the following bitcoin address:





CHANGES.md

# version 0.0.3
## 2023-01-09
- use `entry_points` instead `scripts` in `setup.py` file
- modify `arvan` script to support windows
- now can Create DNS record with type AAAA

---
# version 0.0.2
## 2023-01-05
- now can install via pip
- dynamicly fill of the `setup.py` file with `__version__.py` variables
- add `arvan` script to manage arvancloud cdn and dns

---
# version 0.0.1
## 2022-12-20
- initial release


arvanApi/__main__.py
#!/usr/bin/env python

import logging
import os
from rich.console import Console
from rich.table import Table
#from tabulate import tabulate
from arvanApi import Arvan

logging.basicConfig(level=logging.ERROR, format='%(asctime)s - %(levelname)s - %(message)s')

console = Console()

def clearScreen():
    #console.print("\033c", end="")
    os.system('cls' if os.name == 'nt' else 'clear')


def changeSslStatus(domain, ssl_status):
    if ssl_status :
        console.print("\n> Do you want to disable ssl? (yes/no)")
        ssl = input("> ").strip()
        if ssl in ["yes", "y"] :
            domain.changeSsl(ssl_status=False)
    else :
        console.print("\n> Do you want to enable ssl? (yes/no)")
        ssl = input("> ").strip()
        if ssl in ["yes", "y"] :
            domain.changeSsl(ssl_status=True)


def showDnsRecords(DNSs, showindex=False):
    dnsList = []
    table = Table(show_lines=True, show_header=True, header_style="bold magenta", 
                    row_styles=["dim", ""], highlight=True)
    
    if showindex:
        table.add_column("Index", justify="center")
    table.add_column("Type", justify="center")
    table.add_column("Name", style="bright_cyan", justify="center")
    table.add_column("IP", style="green", justify="center")
    table.add_column("cloud", justify="center")

    index = 0
    for key, rec in DNSs:
        port  = f":{rec.value[0]['port']}" if rec.value[0]['port'] else ""
        host = f"{rec.value[0]['ip']}{port}"
        name = rec.domain if rec.name == '@' else f"{rec.name}.{rec.domain}"
        if showindex:
            table.add_row(str(index), rec.type, name, host, str(rec.cloud))
        else:
            table.add_row(rec.type, name, host, str(rec.cloud))
        dnsList.append([rec.type, name, host, rec.cloud, rec.id])
        index += 1
    # https://stackoverflow.com/questions/9535954
    #table = tabulate(dnsList, headers=['Type', 'name', 'ip', 'cloud', 'id'], tablefmt='fancy_grid', showindex=showindex)
    #console.print(table)

    console.print(table)

    return dnsList


def createDnsRecord(domain):
    # create dns record with type A or AAAA
    console.print("\nCreate DNS record with type A or AAAA")

    console.print("\n> Please enter your dns record name. (subdomain or @ for root domain)")
    dnsName = input("> ").strip()

    console.print("\n> Please enter your dns record ipv4 or ipv6. For example it like (192.168.27.32)")
    dnsIp = input("> ").strip()

    console.print("\n> Please enter your dns record port. For example it like (80)")
    console.print("  port should be in (1, 65535)")
    console.print("  If you don't want to use port, just press enter")
    dnsPort = input("> ")
    try:
        dnsPort = int(dnsPort)
    except:
        dnsPort = None

    console.print("\n> Please enter your dns ttl. For example it like (1800)")
    console.print("  TTL is the time to live for the DNS record. It is the amount of time that the DNS record is cached by the DNS server.")
    console.print("  The lower the TTL, the more often the DNS server will check for changes to the DNS record.")
    console.print("  The higher the TTL, the longer the DNS server will cache the DNS record.")
    console.print("  ttl should be in (120, 180, 300, 600, 900, 1800, 3600, 7200, 18000, 43200, 86400, 172800, 432000)")
    console.print("  If you don't want to use ttl, just press enter")
    dnsTtl = input("> ").strip()
    try:
        dnsTtl = int(dnsTtl)
    except:
        dnsTtl = 432000

    console.print("\n> Do you want to Using the Cloud option? default is no")
    console.print("  the traffic to your server go through Arvan's edge servers so it will be secured and optimized.")
    console.print("  When the Cloud option is disabled, your origin server address is publicly exposed and is not protected against cyber threats.")
    console.print("  If you don't want to use cloud, just press enter. (yes/no)")
    dnsCloud = input("> ").strip()

    console.print("\n> Please enter your Origin server connection protocol.")
    console.print("  If the value is \"default\", the protocol is determined automatically based on the port number.")
    console.print("  If the value is \"auto\", the protocol is determined automatically based on the port number and the certificate of the origin server.")
    console.print("  If the value is \"http\", the protocol is always HTTP.")
    console.print("  If the value is \"https\", the protocol is always HTTPS.")
    console.print("  If you don't want to use upstream_https, just press enter")
    upstream_https = input("> ").strip()
    
    if upstream_https not in ["default", "auto", "http", "https"]:
        upstream_https = 'default'

    if domain.createDnsARecord(
                            dnsName, 
                            dnsIp, 
                            port=dnsPort, 
                            ttl=dnsTtl,
                            cloud=dnsCloud in ["yes", "y"], 
                            upstream_https=upstream_https
                            ) :
        console.print("[+] DNS record created")
    else:
        console.print("[-] DNS record not created")


def deleteDnsRecord(domain):
    console.print("\nDelete DNS record")
    DNSs = showDnsRecords(domain.DNSs.items(), showindex=True)
    if len(DNSs) == 0:
        console.print("[-] No DNS record found")
        return
    
    console.print(f"\n> Please select your dns record [0, {len(DNSs)-1}]")
    dnsIndex = input("> ").strip()
    try:
        dnsIndex = int(dnsIndex)
    except:
        dnsIndex = -1
    if dnsIndex not in range(len(DNSs)):
        console.print("[-] Invalid dns record index")
        return
    
    console.print(f"\n> Are you sure you want to delete this dns record?")
    console.print(f"\t{DNSs[dnsIndex]}")
    console.print("  (yes/no)")
    sure = input("> ").strip()
    if sure not in ["yes", "y"]:
        return
    if domain.deleteDnsById(DNSs[dnsIndex][-1]):
        console.print("[+] DNS record deleted")
    else:
        console.print("[-] DNS record not deleted")


def deleteDomain(arv, domain):
    console.print(f"\n> Are you sure you want to delete this domain?")
    console.print(f"\t{domain.domain}")
    console.print("  (yes/no)")
    sure = input("> ").strip()
    if sure not in ["yes", "y"]:
        return False
    if arv.deleteDomain(domain.domain):
        console.print("[+] Domain deleted")
    else:
        console.print("[-] Domain not deleted")
    return True


def domainMenu(arv, domain):
    while True:
        console.print(f"\nManaging Domain: {domain.domain} \ttype: {domain.type} \tstatus: {domain.status}")
        if domain.current_ns != domain.ns_keys :
            console.print("[X] Change you ns records from your domain provider to the following records")
            console.print(f"\tns keys = {domain.ns_keys}")
            #exit(1)
        ssl_status = domain.getSslSettings()['data']['ssl_status']
        console.print(f"ssl status: {ssl_status}")

        console.print("\nSelect an action:")
        console.print("\t0 - back to main menu")
        console.print("\t1 - change ssl status")
        console.print("\t2 - show dns records")
        console.print("\t3 - add dns record")
        console.print("\t4 - delete dns record")
        console.print("\t5 - delete domain")

        choice = input("> ")
        clearScreen()

        try:
            choice = int(choice)
        except:
            console.print("[-] Invalid choice")
            continue
        
        if choice == 0:
            return
        elif choice == 1:
            changeSslStatus(domain, ssl_status)
        elif choice == 2:
            console.print("\nDNS Records:")
            showDnsRecords(domain.DNSs.items())
        elif choice == 3:
            createDnsRecord(domain)
        elif choice == 4:
            deleteDnsRecord(domain)
        elif choice == 5:
            if deleteDomain(arv, domain):
                return


def startMenu(arv):
    while True:
        domains = arv.getDomains()
        console.print("\nSelect a domain to manage:")
        console.print("\t0 - exit")
        console.print("\t1 - add new domain")
        for i, domain in enumerate(domains):
            console.print(f"\t{i+2} - {domain.domain}")
        
        choice = input("> ")
        clearScreen()

        try:
            choice = int(choice)
        except:
            console.print("[-] Invalid choice")
            continue
        
        if choice == 0:
            console.print("\nGoodbye!")
            break
        elif choice == 1:
            console.print("\n> Please enter your domain name. For example it like (example.com)")
            domainName = input("> ").strip()
            if arv.createDomain(domainName):
                console.print("[+] Domain created")
            else:
                console.print("[-] Domain not created")
        elif choice < len(domains)+2:
            domain = domains[choice-2]
            domainMenu(arv, domain)
        else:
            console.print("[-] Invalid choice")


def main() -> None:
    console.print("> Welcome to Arvan CDN Manager")
    console.print("> Please enter your Arvan ApiKey. For example it like (Apikey df0bbd86-8f94-5c1e-a6ebasdasd-asdasd)")
    console.print("  If you don't have an ApiKey, you can get it from https://panel.arvancloud.ir/profile/machine-user")
    apikey = input("> ").strip()

    console.print("")
    with console.status("[bold cyan]Checking ApiKey...") as status:
        arv = Arvan(apikey, debug=False)
        if arv._domainsDict == False :
            exit(1)
        console.print("[bold green]> ApiKey authenticated")
    
    startMenu(arv)


if __name__ == "__main__":
    main()



arvanApi/arvanModule.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Doc @ https://www.arvancloud.ir/docs/api/cdn/4.0
# Api Key @ https://panel.arvancloud.ir/profile/api-keys
import json
import logging
from time import time
import requests


class ArvanDNS:
    def __init__(self, data, domain, debug=False):
        # for key in data:
        #     setattr(self, key, data[key])
        self.domain = domain
        self.id = data.get('id', None)
        self.type = data.get('type', None)
        self.name = data.get('name', None)
        self.value = data.get('value', None)
        self.ttl = data.get('ttl', None)
        self.cloud = data.get('cloud', None)
        self.upstream_https = data.get('upstream_https', None)
        self.ip_filter_mode = data.get('ip_filter_mode', None)
        self.can_delete = data.get('can_delete', None)
        self.is_protected = data.get('is_protected', None)
        self.health_check_status = data.get('health_check_status', None)
        self.health_check_setting = data.get('health_check_setting', None)
        self.created_at = data.get('created_at', None)
        self.updated_at = data.get('updated_at', None)
        self.monitoring_status = data.get('monitoring_status', None)
        self.health_check = data.get('health_check', None)

    def __repr__(self) -> str:
        if self.type == 'a':
            port  = f":{self.value[0]['port']}" if self.value[0]['port'] else ""
            if self.name == '@':
                return f"[{self.type}] {self.domain} -> {self.value[0]['ip']}{port}\t ID: {self.id}"
            else:
                return f"[{self.type}] {self.name}.{self.domain}  -> {self.value[0]['ip']}{port}\t ID: {self.id}"
        else:
            return f"[{self.type}] {self.name}\t ID: {self.id}"


class ArvanDomain:
    def __init__(self, api_key, data, debug=False):
        self.API_KEY = f"{api_key}"
        self.HEADERS = {
                    'Authorization': f'{self.API_KEY}',
                    'Content-Type': 'application/json'
                }
        
        self.id = data.get('id', None)
        self.user_id = data.get('user_id', None)
        self.domain = data.get('domain', None)
        self.name = data.get('name', None)
        self.services = data.get('services', {})
        self.dns_cloud = data.get('dns_cloud', None)
        self.plan_level = data.get('plan_level', None)
        self.features = data.get('features', None)
        self.smart_routing_status = data.get('smart_routing_status', None)
        self.ns_keys = data.get('ns_keys', [])
        self.current_ns = data.get('current_ns', [])
        self.status = data.get('status', None)
        self.parent_domain = data.get('parent_domain', None)
        self.is_paused = data.get('is_paused', None)
        self.is_suspended = data.get('is_suspended', None)
        self.type = data.get('type', None)
        self.cname_target = data.get('cname_target', None)
        self.custom_cname = data.get('custom_cname', None)
        self.transfer = data.get('transfer', None)
        self.created_at = data.get('created_at', None)
        self.updated_at = data.get('updated_at', None)
        self.DEBUG = debug
        self.BASE_URL = f"https://napi.arvancloud.ir/cdn/4.0/domains/{self.domain}"
        self.DNSs = self._getDNS()

    def _getDNS(self, search='', page=1, rType=None, per_page=300):
        url = f'{self.BASE_URL}/dns-records'
        params = {
                    'search':   f'{search}',
                    'page':     f'{page}',
                    'type':     f'{rType}',
                    'per_page': f'{per_page}'
                }
        r = requests.get(url, params=params, headers=self.HEADERS)
        DnsList ={}
        if r.status_code == 200 :
            res = r.json()
            for record in res['data']:
                rec = ArvanDNS(record, self.domain, self.DEBUG)
                if rec.type == 'a':
                    DnsList[(rec.type, rec.name)] = rec
                if self.DEBUG:
                    logging.debug(f"\t{rec}")
            if res['links']['next'] :
                DnsList.update(self._getDNS(search, page+1, rType, per_page))
        elif r.status_code == 401:
            logging.error("[_getDNS] Access token is missing or invalid")
        elif r.status_code == 404:
            logging.error("[_getDNS] Domain not found")
        else:
            logging.error(f"[_getDNS] Err with status code {r.status_code}")
        return DnsList

    def getDnsRecords(self, type, name):
        return self.DNSs.get((type, name), None)

    def createDnsARecord(self, name, ipAddr, port=None, ttl=432000, cloud=False, upstream_https='default'):
        url = f'{self.BASE_URL}/dns-records'

        acceptTtl = [120, 180, 300, 600, 900, 1800, 3600, 7200, 18000, 43200, 86400, 172800, 432000]
        ttl = ttl if ttl in acceptTtl else 432000

        payload = {
                    'type': 'a',
                    'name': f'{name}',
                    'value':[
                        {
                            'ip': f'{ipAddr}',
                            'port': port
                        }
                    ],
                    'ttl':ttl,
                    'cloud': cloud,
                    'upstream_https': f'{upstream_https}'
                }

        r = requests.post(url, data=json.dumps(payload), headers=self.HEADERS )

        if r.status_code == 201 :
            rec = ArvanDNS(r.json()['data'], self.domain, self.DEBUG)
            self.DNSs[(rec.type, rec.name)] = rec
            logging.info(f"DNS record created: {rec}")
            return r.json()['data']['id']
        elif r.status_code == 401:
            logging.error("[createDnsARecord] Access token is missing or invalid")
        elif r.status_code == 404:
            logging.error("[createDnsARecord] Resource not found")
        elif r.status_code == 422:
            logging.error("[createDnsARecord] The given data was invalid")
        else:
            logging.error(f"[createDnsARecord] Err with status code {r.status_code}")
        return None
    
    def deleteDnsByName(self, name, type='a'):
        if (type, name) not in self.DNSs:
            logging.error(f"[deleteDnsByName] DNS record {name} not found")
            return False
        
        rId = self.getDnsRecords(type, name).id
        url = f"{self.BASE_URL}/dns-records/{rId}"
        r = requests.delete(url, headers=self.HEADERS )

        if r.status_code == 200 :
            logging.info(f"[delDnsRecord] DNS record deleted") 
            del self.DNSs[(type, name)]
            return True
        elif r.status_code == 401:
            logging.error("[delDnsRecord] Access token is missing or invalid")
        elif r.status_code == 404:
            logging.error("[delDnsRecord] Resource not found")
        else:
            logging.error(f"[delDnsRecord] Err with status code {r.status_code}")
        return False
    
    def deleteDnsById(self, rId):
        url = f"{self.BASE_URL}/dns-records/{rId}"
        r = requests.delete(url, headers=self.HEADERS )

        if r.status_code == 200 :
            logging.info(f"[delDnsRecord] DNS record deleted")
            self.DNSs = self._getDNS()
            return True
        elif r.status_code == 401:
            logging.error("[delDnsRecord] Access token is missing or invalid")
        elif r.status_code == 404:
            logging.error("[delDnsRecord] Resource not found")
        else:
            logging.error(f"[delDnsRecord] Err with status code {r.status_code}")
        return False
    
    def getSslSettings(self):
        url = f'{self.BASE_URL}/ssl'
        r = requests.get(url, headers=self.HEADERS)
        if r.status_code == 200 :
            logging.info(f"Succesfully retrieved SSL settings for {self.domain}")
            return r.json()
        elif r.status_code == 401:
            logging.error("Access token is missing or invalid")
        elif r.status_code == 404:
            logging.error("Resource not found")
        return None
    
    def changeSsl(self, ssl_status=True): # True for enable, False for disable
        url = f'{self.BASE_URL}/ssl'
        payload = { 'ssl_status': ssl_status }
        r = requests.patch(url, data=json.dumps(payload), headers=self.HEADERS  )
        if r.status_code == 200 :
            logging.info(f"Succesfully changed SSL settings for {self.domain}")
            return True
        elif r.status_code == 401:
            logging.error("Access token is missing or invalid")
        elif r.status_code == 404:
            logging.error("Resource not found")
        else:
            logging.error(f"Err with status code {r.status_code}")
        return False

    def _deleteDomain(self):
        url = f'{self.BASE_URL}'
        payload = { 'id': self.id }
        r = requests.delete(url, data=json.dumps(payload), headers=self.HEADERS  )
        if r.status_code//100 == 2:
            logging.info(f"Domain {self.domain} deleted")
            return True
        elif r.status_code == 401:
            logging.error("Access token is missing or invalid")
        elif r.status_code == 404:
            logging.error("Resource not found")
        elif r.status_code == 422:
            logging.error(f"The given data was invalid")
        else:
            logging.error(f"Err with status code {r.status_code}")
        return False
        
    def __repr__(self) -> str:
        return f"Domain: {self.domain} \t Type: {self.type} \t ID: {self.id}"


class Arvan:
    def __init__(self, api_key, debug=False):
        self.BASE_URL = "https://napi.arvancloud.ir/cdn/4.0"
        self.API_KEY = api_key if "Apikey" in api_key else f"Apikey {api_key}" 
        self.HEADERS = {
                    'Authorization': f'{self.API_KEY}',
                    'Content-Type': 'application/json'
                }
        self.DEBUG = debug
        self._domainsDict = self.initDomains()

    def initDomains(self, search='', page=1, per_page=300):
        url = f'{self.BASE_URL}/domains'
        params = {
                    'search':   f'{search}',
                    'page':        page,
                    'per_page':    per_page
                }
        r = requests.get(url, params=params, headers=self.HEADERS )
        DomainList = {}
        if r.status_code == 200 :
            res = r.json()
            for record in res['data']:
                rec = ArvanDomain(self.API_KEY, record, debug=self.DEBUG)
                DomainList[record['domain']] = rec
                if self.DEBUG :
                    logging.debug(f"{rec}")
            if res['links']['next'] :
                DomainList.update(self.GetDomains(search, page+1, per_page))
            return DomainList
        elif r.status_code == 401:
            logging.error("Access token is missing or invalid")
        else:
            logging.error(f"Err with status code {r.status_code}")
        return False

    def createDomain(self, domain):
        if domain in self._domainsDict:
            logging.error(f"Domain {domain} already exists")
            return None
        
        url = f"{self.BASE_URL}/domains/dns-service"
        payload = {
                    'domain': f'{domain}',
                    'domain_type':'full',
                    'plan_level': 1,
                }
        r = requests.post(url, data=json.dumps(payload), headers=self.HEADERS)
        if r.status_code//100 == 2 :
            rec = ArvanDomain(self.API_KEY, r.json()['data'], debug=self.DEBUG)
            self._domainsDict[domain] = rec
            if self.DEBUG :
                logging.debug(f"New domain {domain} created!")
                logging.debug(f"please set ns to {rec.ns_keys}")
            return True
        elif r.status_code == 401:
            logging.error("Access token is missing or invalid")
        elif r.status_code == 422:
            logging.error(f"The given data was invalid")
        else:
            logging.error(f"Err with status code {r.status_code}")
        return False

    def deleteDomain(self, domain):
        if domain in self._domainsDict:
            if self._domainsDict[domain]._deleteDomain():
                del self._domainsDict[domain]
                return True
        return False
    
    def getDomains(self):
        return list(self._domainsDict.values())

    def getDomain(self, Domain):
        return self._domainsDict.get(Domain, None)
    
    def __repr__(self) -> str:
        return f"ArvanApi with api_key: {self.API_KEY}"

